File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int per = 10007;
int visited[200005]={0,};
int dp[200005]={0,};

int solution(int n, vector<int> tops) {
int answer = 0;
for(int i=0; i<tops.size(); i++){
	if(tops[i] == 1){
		visited[i*2 +1] = 1;
	}
}

dp[0] =1;
dp[1] = 2;

if(visited[1] ==1){
	dp[1] = 3;
}

for(int j =2; j<=2*n; j++){
	// index가홀수이고 위 삼각형이 있을때.
	if(j%2 == 1 && visited[j] == 1){ 
		dp[j] = dp[j-1] + dp[j-1] + dp[j-2];
	}else{
	
    //짝수
	dp[j] = dp[j-1] + dp[j-2];
	}
	dp[j] %= 10007;
}

answer = dp[2*n];
return answer;
}

File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.cs
using System;

public class Solution {
    public int solution(int n, int[] tops) {
        int[,] dp = new int[100001, 2];
        dp[0,0] = 2 + tops[0];
        dp[0,1] = 1;
        
        for(int i = 1; i < tops.Length; i++){
            dp[i,0] = dp[i-1,0] * (2 + tops[i]) + dp[i-1,1] * ( 1+ tops[i]);
            dp[i,1] = dp[i-1,0] + dp[i-1,1];
            
            dp[i,0] %= 10007;
            dp[i,1] %= 10007;
        }
        int answer = (dp[n-1,0] + dp[n-1,1]) % 10007;
        return answer;
    }
}